Add preserve_original setting in ngram token filter

amitmbm opened the pull request.

amitmbm committed:
Completed TODO task of adding preserve_original setting in ngram-token-filter and added the missing test class.

elasticmachine commented:
Pinging @elastic/es-search (:Search/Analysis)

markharwood commented:
Thanks for this contribution, @amitmbm
 
A couple of extra things to consider - 
1) Adding a test [here](https://github.com/elastic/elasticsearch/blob/master/modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/NGramTokenizerFactoryTests.java)
2) Documenting the parameter [here](https://github.com/elastic/elasticsearch/blob/master/docs/reference/analysis/tokenfilters/ngram-tokenfilter.asciidoc#configurable-parameters)

amitmbm commented:
@markharwood  thanks for the suggestions, I'll address them soon and let you know.

markharwood commented:
>Adding a test here

My bad - that's the wrong test class - it's for the NGramTokenizer rather than the NGramTokenFilter you modified. I'm not sure there is an existing class for testing that TokenFilter

amitmbm commented:
@markharwood Sorry I forgot, yeah yesterday I searched for the test class of this but couldn't find one and should have mentioned that. 

Also, I added the documentation of newly added `preserve_original` param. 

Now please let me know if I need to do something else, would like to close this before I go to bed today :)

amitmbm commented:
@markharwood Hope you are doing well. Please go through my last commit and let me know if I need to address some more comments.

markharwood commented on docs/reference/analysis/tokenfilters/ngram-tokenfilter.asciidoc:
```diff
@@ -185,6 +185,10 @@ Maximum length of characters in a gram. Defaults to `2`.
+`preserve_original`::
+(Optional, boolean)
+If set to true then it would also emit the original token. Defaults to `true`.
+
```
The default is false? (And needs to be for backwards-compatibility reasons).

amitmbm replied:
@markharwood thanks for pointing this, Addressed it in my lastest comment.

amitmbm replied:
@markharwood I also rebased my feature branch with the latest master code and squashed my commits. Let me know if anything else required from my side.

amitmbm commented:
@markharwood Added the missing test file `NGramTokenFilterFactoryTests`  and ran the test locally and these are passing. and the following is the output:

> A build scan will not be published due to this build running offline.
> 1:21:23 PM: Tasks execution finished ':modules:analysis-common:cleanTest :modules:analysis-common:test --tests "org.elasticsearch.analysis.common.NGramTokenFilterFactoryTests"'.

amitmbm commented:
@elasticmachine ok to test

markharwood commented:
Hi Amit.
Apologies for not replying sooner - I've been tied up on adding a different feature..
I see Christoph has helped in the meantime and another PR has been merged. 
Can I close this now?
Thanks very much for your contribution!

amitmbm commented:
@markharwood It's OK I can understand. Christoph helped me with similar PR in `edge-ngram token-filter` while this PR is for `ngram token filter`, it also needs to be merged.

Could you please do the final review, please note that I've added a missing test class as well, as pointed by you earlier and addressed all other comments already.

markharwood commented:
Many thanks for this, Amit.
I've merged to master (8.0) and backported to 7.x (7.8) branches.
